### ISSUE TITLE
Feature request: Pre-configured store hooks (object, array, etc …)

### ISSUE THREAD

**Issue by tengl** [mausworks/tyin#8]
Hi there,

I will try this lib out on an internal tool. It is a bit verbose when it comes to creating and configuring stores, but that is fine, you can encapsulate the code. 

Wouldn't it be nice though with some pre-configured store hooks with the most common usages? E.g.

```ts
import storeHook, { StateSelectorHook } from "tyin/hook";
import extend, { Extensible } from "tyin/extend";
import { StoreAPI, StoreOptions } from "tyin/store";
import objectAPI, { ObjectAPI, ObjectLike } from "tyin/plugin-object";

export function objectStoreHook<T extends ObjectLike | null>(
  initialState: T,
  options?: StoreOptions<T>,
  sealed?: true
): StateSelectorHook<T> & StoreAPI<T> & ObjectAPI<T>;
export function objectStoreHook<T extends ObjectLike | null>(
  initialState: T,
  options?: StoreOptions<T>,
  sealed?: false
): Extensible<StateSelectorHook<T> & StoreAPI<T> & ObjectAPI<T>>;
export function objectStoreHook<T extends ObjectLike | null>(
  initialState: T,
  options?: StoreOptions<T>,
  sealed: boolean = true
):
  | (StateSelectorHook<T> & StoreAPI<T> & ObjectAPI<T>)
  | Extensible<StateSelectorHook<T> & StoreAPI<T> & ObjectAPI<T>> {
  const hook = extend(storeHook(initialState, options)).with(objectAPI());
  if (sealed) return hook.seal();
  return hook;
}
```

(The extra function definitions is to handle if it is sealed or not, there is probably a better way to do it)

It would make it easy to use:

```ts
const useUserState = objectStoreHook<User | null>(null);
```

The same can be done for an array store hook. 

Also, I noticed that the `objectAPI` doesn't allow `undefined`. Is that by design?

**Comment by mausworks** [comment]
I'm really happy to hear that you want to give this library a go!

This is one of the things that I was struggling with while doing the initial design of Tyin. I first did three separate stores (vanilla, object, array), as you suggested, but when migrating dott.bio from zustand to Tyin, I noticed that the extra imports generally didn't bother me, <kbd>CMD</kbd> + <kbd>.</kbd> (autoimport) generally helps a lot. Most stores ended up producing <10 lines of code, even with the imports—which is A-OK with me.-

I also reasoned that, with the [reducer plugin](https://github.com/mausworks/tyin/pull/7) coming up, I don't want to give preference to "some plugins". Should there be a `reducerStoreHook` too? what about a `persistedStoreHook`? With #9 in mind, what about a `loadableStoreHook`?  I don't mean this in a snide way, I just don't know where to draw the line right now. There is likely some "middle way" here, but I'm still working on seeing it!

**Comment by tengl** [comment]
Yeah, I guess it isn't much code that is written, but it looks nicer and cleaner when it is used. 

In a bigger project I would probably make some pre-configured stores that are commonly used directly in the application code. 

I guess you could have a hooks module for each plugin, e.g. `tyin/plugin-object-hooks`. Or maybe you should create a new `tyin-react` project that imports `tyin` and adds React hooks support with hooks for the plugins as well (where applicable). You will get rid of the `react` dependency as well in `tyin` and the examples could be more targeted to the usage with React.
